### PR TITLE
Implement root scanning for garbage collected objects

### PR DESCRIPTION
**Pull request by jacob-hughes**
_[It may make reviewing a lot easier to look at this PR in chronological commit order; the diff for  f4f83d6 is quite large, but it simply hides some things which were public when they shouldn't have been and creates a single API to the alloc metadata which can be used by the GC.]_

Users can trigger a single collection through a public `collect()` function, which, at present, simply marks objects reachable directly from the root-set. Reclamation of unreachable GC objects is not yet
implemented.

We define the root-set as any reference live on the stack during a collection or values inside callee-save registers. Since register spilling is inherently platform specific, we require assembly stubs containing the spill code for each architecture. At present, we only implement X86_64.

A secondary debug API exists for helping query information about a collection for tests etc.

The collector is **not** in anyway thread-safe (yet).

There's a fair bit of assembly magic here, so I think it might be wise to cc @vext01 on this one too.

### CODE REVIEW

File: src/SpillRegisters_X64.S [hunk 1]
```diff
@@ -0,0 +1,27 @@
+.text
+
+.extern scan_stack
+.globl spill_registers
+
+spill_registers:
+    // Push all callee-saves registers to get them on the stack for conservative
```
Comment (vext01) [inline src/SpillRegisters_X64.S]: Typo: "callee-save"

File: src/SpillRegisters_X64.S [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    // stack scanning. We maintain 16-byte alignment at calls (Sys V ABI). There
+    // is an 8-byte return address on the stack and we push 56 bytes which
+    // maintains 16-byte stack alignment at the call.
+    push $0
```
Comment (vext01) [inline src/SpillRegisters_X64.S]: What's `$0` again?

File: src/SpillRegisters_X64.S (hunk 1 above)
Comment (vext01) [reply src/SpillRegisters_X64.S]: Did you want to document why it's only the callee save registers that we care about?

File: src/SpillRegisters_X64.S (hunk 2 above)
Comment (jacob-hughes) [reply src/SpillRegisters_X64.S]: GAS syntax requires all constants to be prefixed with `$`. We're simply pushing padding to the stack to maintain 16-byte alignment.

[As an aside, I ran into a nasty gotcha with this one. I'd written `push 0` before, which actually pushes the contents of memory address 0, which, of course, segfaults as it tries to dereference it. :man_facepalming:  ]

File: src/SpillRegisters_X64.S (hunk 1 above)
Comment (jacob-hughes) [reply src/SpillRegisters_X64.S]: Sure, done.

File: src/SpillRegisters_X64.S (hunk 2 above)
Comment (vext01) [reply src/SpillRegisters_X64.S]: Ah yes. Good old GNU syntax.

File: src/SpillRegisters_X64.S (hunk 1 above)
Comment (vext01) [reply src/SpillRegisters_X64.S]: LGTM

File: src/alloc.rs [hunk 3]
```diff
@@ -382,6 +372,58 @@ fn exit_alloc(msg: &str) {
...
+                None => {
+                    let mut al = AllocList::new();
+                    al.insert(ptr as usize, size, gc);
+                    ALLOC_INFO = Some(al);
+                }
+            };
+        }
+        ALLOC_LOCK.unlock();
+    }
+    /// Returns metadata about an allocation block when given an arbitrary
```
Comment (ltratt) [inline src/alloc.rs]: Since this function returns `Option`, I assume you can pass *any* pointer in here? The doc string suggests that only valid pointers can be passed in but I don't think that's correct.

Comment (jacob-hughes) [reply src/alloc.rs]: Yes, I see what you mean by the wording of the docstring. I'll reword it to make it clearer that you _can_ pass any pointer. But it will only return pointer info in those circumstances.

File: src/alloc.rs [hunk 4]
```diff
@@ -382,6 +372,58 @@ fn exit_alloc(msg: &str) {
...
+        ALLOC_LOCK.unlock();
+        ptr_info
+    }
+
+    /// Removes the metadata associated with an allocation.
+    pub(crate) fn remove(ptr: usize) {
+        ALLOC_LOCK.lock();
+        let ptr_info = unsafe { ALLOC_INFO.as_ref().unwrap().find_base(ptr).unwrap() };
+
+        debug_assert!(ptr_info.gc); // FIXME: Move this expensive function into the assertion
```
Comment (ltratt) [inline src/alloc.rs]: I don't understand the assertion: the `assert` isn't a function call. You can, from memory, do things like:

```rust
#[cfg(debug)]
{
    let ptr_info = ...;
    assert!(ptr_info.gc);
}
```

if that's what you wanted?

File: src/lib.rs [hunk 5]
```diff
@@ -101,12 +124,38 @@ impl<T> Gc<T> {
...
     }
 }
 
+impl GcHeader {
+    pub(crate) fn new() -> Self {
+        Self(0)
+    }
+
+    pub(crate) fn mark_bit(&self) -> bool {
+        ((self.0 >> 0) & 1) == 1
```
Comment (ltratt) [inline src/lib.rs]: Isn't `n >> 0` simply `n`?

File: src/alloc.rs (hunk 4 above)
Comment (jacob-hughes) [reply src/alloc.rs]: Sorry, that comment shouldn't have ended up in the PR and was a reminder for me to do exactly what you suggested.

File: src/gc.rs [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+use crate::{
+    alloc::{AllocMetadata, PtrInfo},
+    Gc,
+};
+use std::sync::Mutex;
+
+static WORD_SIZE: usize = 8; // Bytes
```
Comment (ltratt) [inline src/gc.rs]: I suggest you guard this with `#[cfg(target_pointer_width = "64")]` so that if anyone tries compiling this on <>64 bit machine they'll get a static error.

File: src/gc.rs [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+
+type StackAddress = usize;
+type StackValue = usize;
+
+type StackScanCallback = extern "sysv64" fn(&mut Collector, StackAddress);
+#[link(name = "SpillRegisters", kind = "static")]
+extern "sysv64" {
+    // Rust (correctly) warns us that `Collector` has an unspecified layout and
+    // that it is UB to pass this to C code via FFI. However, our register spill
+    // code does not dereferences this, nor does it ever move the `Collector`
```
Comment (ltratt) [inline src/gc.rs]: "dereference"s -> "dereference".

File: src/gc.rs [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+static WORD_SIZE: usize = 8; // Bytes
+
+type StackAddress = usize;
+type StackValue = usize;
+
+type StackScanCallback = extern "sysv64" fn(&mut Collector, StackAddress);
+#[link(name = "SpillRegisters", kind = "static")]
+extern "sysv64" {
+    // Rust (correctly) warns us that `Collector` has an unspecified layout and
+    // that it is UB to pass this to C code via FFI. However, our register spill
```
Comment (ltratt) [inline src/gc.rs]: Why don't we pass it as, say, a `const *u8` which will probably get rid of the warning?

File: src/gc.rs [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+pub struct DebugFlags {
+    mark_only: bool,
+}
+
+impl DebugFlags {
+    pub fn new() -> Self {
+        Self { mark_only: false }
+    }
+
+    pub fn none() -> Self {
```
Comment (ltratt) [inline src/gc.rs]: What's the point of this function?

File: src/gc.rs [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+
+    pub(crate) fn collect(&mut self) {
+        // First check that no call to collect is active
+        {
+            let mut cstate = self.state.lock().unwrap();
+            match *cstate {
+                CollectorState::Ready => *cstate = CollectorState::Marking,
+                _ => {
+                    // Bail out if we reach this point as the collector has already
+                    // been invoked -- and is running -- on another thread.
```
Comment (ltratt) [inline src/gc.rs]: Simply saying "The collector is running in another thread" is probably good enough here?

File: src/lib.rs [hunk 11]
```diff
@@ -156,12 +165,53 @@ impl<T> Deref for Gc<T> {
...
 
+pub fn init(flags: gc::DebugFlags) {
+    unsafe { COLLECTOR = Some(Collector::new(flags)) };
+}
+
+pub fn collect() {
+    // FIXME: Is this UB? Theoretically this pub `collect` fn can be called in
+    // parallel from multiple threads, allowing us potentially multiple mutable
+    // references to the underlying `COLLECTOR` struct. However, the inner
+    // `collect` method guards against multiple collection invocations via a
```
Comment (ltratt) [inline src/lib.rs]: Why would this be UB? Merely having a reference to a struct in multiple threads cannot lead to UB (unless the underlying memory moves or similar!): it's reading and writing that one has to be careful about.

File: src/lib.rs [hunk 12]
```diff
@@ -156,12 +165,53 @@ impl<T> Deref for Gc<T> {
...
+}
+
+pub struct Debug;
+
+impl Debug {
+    pub fn is_black<T>(gc: Gc<T>) -> bool {
+        let collector = unsafe { COLLECTOR.as_ref().unwrap() };
+        let cstate = *collector.state.lock().unwrap();
+
+        // Checking an object colour only makes sense immediately after marking has
```
Comment (ltratt) [inline src/lib.rs]: "object colour" -> "object's colour"

File: src/lib.rs [hunk 13]
```diff
@@ -156,12 +165,53 @@ impl<T> Deref for Gc<T> {
...
+        // taken place. After a full collection has happened, marking results are
+        // stale and the object graph must be re-marked in order for this query
+        // to be meaningful.
+        assert_eq!(cstate, gc::CollectorState::FinishedMarking);
+        return collector.colour(unsafe { Gc::from_raw(gc.objptr as *const i8) })
+            == gc::Colour::Black;
+    }
+
+    /// This should be used with caution. For example, if writing a test using
```
Comment (ltratt) [inline src/lib.rs]: This doc string feels a bit confusing. Why not just say something like "If true, this object was not marked for collection in the last sweep"? [BTW, in a sense, your whole warning also applies to `is_black`, since the two functions are simply inverses.]

Comment (jacob-hughes) [reply src/lib.rs]: I've reworded this.

I think the warning isn't as relevant to `is_black` as the conservative GC will tend to over-approximate - if I assert an object should be black after a collection and it's not, that's a bug! 

On the other hand, I can't be sure that an object didn't get marked, as for some non-deterministic reason, residual pointers may have been left lingering on the stack or within registers.

File: src/lib.rs (hunk 12 above)
Comment (jacob-hughes) [reply src/lib.rs]: Fixed

File: src/lib.rs (hunk 11 above)
Comment (jacob-hughes) [reply src/lib.rs]: Makes sense. I've removed the comment.

File: src/gc.rs (hunk 10 above)
Comment (jacob-hughes) [reply src/gc.rs]: Fixed

File: src/gc.rs (hunk 9 above)
Comment (jacob-hughes) [reply src/gc.rs]: Premature over-engineering. It's gone

File: src/gc.rs (hunk 8 above)
Comment (jacob-hughes) [reply src/gc.rs]: Okay

File: src/gc.rs (hunk 7 above)
Comment (jacob-hughes) [reply src/gc.rs]: Fixed

File: src/gc.rs (hunk 6 above)
Comment (jacob-hughes) [reply src/gc.rs]: I've decided to guard the entire library, as there are other places where this assumption is made too.

File: src/gc.rs [hunk 14]
```diff
@@ -0,0 +1,229 @@
+// Copyright (c) 2019 King's College London
```
[the remark this answers is not in the document]
Comment (jacob-hughes) [reply src/gc.rs]: Okay done. For future reference, should the main `lib.rs` / `mod.rs` file only contain the full license?

File: src/lib.rs (hunk 5 above)
Comment (jacob-hughes) [reply src/lib.rs]: Yes. Fixed

File: src/lib.rs (hunk 13 above)
Comment (ltratt) [reply src/lib.rs]: I think none of this is a problem if you avoid talking about e.g "liveness" (so, for example, it would definitely have been incorrect to say "this guarantees the object was live at the last collection). That said, what can anyone do with such a guarantee? I think the answer is "nothing": by definition, the instant the collection is finished, this information is stale and can't be relied on. So I think you might be worrying unduly :)

File: src/gc.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (ltratt) [reply src/gc.rs]: What we've gone for, mostly, is a `LICENSE` file and then a short header at the top of each file e.g. see `Vob` https://github.com/softdevteam/vob/blob/master/src/lib.rs

File: src/alloc.rs [hunk 15]
```diff
@@ -413,9 +413,11 @@ impl AllocMetadata {
     /// Removes the metadata associated with an allocation.
     pub(crate) fn remove(ptr: usize) {
         ALLOC_LOCK.lock();
-        let ptr_info = unsafe { ALLOC_INFO.as_ref().unwrap().find_base(ptr).unwrap() };
 
-        debug_assert!(ptr_info.gc); // FIXME: Move this expensive function into the assertion
+        debug_assert!(|| -> bool {
```
Comment (ltratt) [inline src/alloc.rs]: Ha, that is not a way of doing it that had occurred to me!

File: src/gc.rs [hunk 16]
```diff
@@ -21,15 +21,10 @@ type StackValue = usize;
...
-    // instance by value. It simply moves the address to the first argument
-    // register to be used as the implicit `self` address in the callback. It is
-    // therefore safe to do this without specifying a C style layout for the
-    // struct.
-    #[allow(improper_ctypes)]
-    fn spill_registers(collector: &mut Collector, callback: StackScanCallback);
+    // Pass a type-punned pointer to the collector and move it to the asm spill
+    // code. This is so it can be passed straight back as the implicit `self`
+    // address in the callback.
+    fn spill_registers(collector: *mut i8, callback: StackScanCallback);
```
Comment (ltratt) [inline src/gc.rs]: I think it's more conventional to use `*mut u8`, but I might be wrong.

File: src/lib.rs (hunk 13 above)
Comment (jacob-hughes) [reply src/lib.rs]: Sure. I suppose the main point that I am trying to make it clear is that it's safe to use `is_black` in a test, but not `is_white`. It's always safe to over-approximate, and it doesn't really matter if a white object was coloured black as a false positive. However, we can't run a test collection and assert that objects which we didn't expect to be coloured black remain white.

File: src/gc.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (jacob-hughes) [reply src/gc.rs]: Okay, so I should really replace the license with the shorter form in the other files too?

File: src/gc.rs (hunk 16 above)
Comment (jacob-hughes) [reply src/gc.rs]: Sure, I think I probably meant to write `u8` anyway.

**Comment by ltratt** [comment]
@jacob-hughes I assume you meant to remind me to rereview this?

File: src/gc.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (ltratt) [reply src/gc.rs]: Yeah. You can do it in another PR if that makes more sense from your perspective.

File: src/lib.rs (hunk 13 above)
Comment (jacob-hughes) [reply src/lib.rs]: > Unless, that is, one explicitly says "you can not safely use !o.is_black()"?

Exactly `!is_black()` is not something you should ever test for. Perhaps I should make that clearer in the doc comments. It's only right to test that a collector reached reachable objects in the last collection.

**Comment by jacob-hughes** [comment]
Yes sorry

File: src/alloc.rs [hunk 17]
```diff
@@ -116,16 +116,6 @@ unsafe impl Alloc for GCMalloc {
     unsafe fn alloc(&mut self, layout: Layout) -> Result<NonNull<u8>, AllocErr> {
         let ptr = libc::malloc(layout.size() as libc::size_t) as *mut u8;
 
-        ALLOC_LOCK.lock();
```
Comment (vext01) [inline src/alloc.rs]: So how comes you no longer record the allocations here?

File: src/lib.rs [hunk 18]
```diff
@@ -98,6 +98,8 @@ impl<T> Gc<T> {
         unsafe {
             let baseptr = GC_ALLOCATOR.alloc(layout).unwrap().as_ptr();
             let objptr = baseptr.add(uoff);
+
+            AllocMetadata::insert(objptr as usize, layout.size(), true);
```
Comment (vext01) [inline src/lib.rs]: I see, so you do the insertion here...

**Comment by vext01** [comment]
Typo in the first commit message "...to get interact with allocation block metadata". Get or interact?

File: src/lib.rs (hunk 18 above)
Comment (jacob-hughes) [reply src/lib.rs]: Yes, and it's subtly different. We insert the `objptr` for GC objects, not the `baseptr`. This means that theoretically, if there existed only a pointer on the stack to a GC object's `baseptr`, it would not be considered reachable and would be reclaimed.

File: src/lib.rs [hunk 19]
```diff
@@ -101,12 +124,38 @@ impl<T> Gc<T> {
...
+        Self(0)
+    }
+
+    pub(crate) fn mark_bit(&self) -> bool {
+        ((self.0 >> 0) & 1) == 1
+    }
+
+    pub(crate) fn set_mark_bit(&mut self, mark: bool) {
+        if mark {
+            self.0 |= 1 << 0
```
Comment (vext01) [inline src/lib.rs]: More zero shifts here.

File: src/lib.rs [hunk 20]
```diff
@@ -101,12 +124,38 @@ impl<T> Gc<T> {
...
+        } else {
+            self.0 &= !(1 << 0)
```
Comment (vext01) [inline src/lib.rs]: I think it was naughty of Rust to use `!` for bitwise compliment on integers instead of `~`.

Comment (jacob-hughes) [reply src/lib.rs]: I agree

File: src/lib.rs [hunk 21]
```diff
@@ -101,12 +124,38 @@ impl<T> Gc<T> {
...
+        }
+    }
+}
+
+impl<T> Deref for Gc<T> {
+    type Target = T;
+
+    fn deref(&self) -> &T {
+        unsafe { &*(self.objptr as *const T) }
```
Comment (vext01) [inline src/lib.rs]: What's unsafe here? `&*`?

File: gc_tests/tests/on_stack_root_marking.rs [hunk 22]
```diff
@@ -0,0 +1,24 @@
...
+use gcmalloc::{collect, Debug };
+use gcmalloc::gc::DebugFlags;
+
+fn foo() {
+    let y = Gc::new(456 as usize);
+    collect();
+    assert!(Debug::is_black(y));
+}
+
+fn main() {
```
Comment (vext01) [inline gc_tests/tests/on_stack_root_marking.rs]: quite a lot of whitespace here.

File: src/lib.rs (hunk 21 above)
Comment (jacob-hughes) [reply src/lib.rs]: Yes, it's de-referencing a raw pointer.

File: src/gc.rs (hunk 10 above)
Comment (vext01) [reply src/gc.rs]: Laurie an I agreed to exhaustively enumerate (without `_`) in cases where it is practical, as it forces us to consider behaviours in the case that we add another variant.

File: src/gc.rs [hunk 23]
```diff
@@ -0,0 +1,229 @@
...
+                    break;
+                }
+
+                self.mark(obj, Colour::Black);
+            }
+        }
+    }
+
+    #[no_mangle]
+    extern "sysv64" fn scan_stack(&mut self, rsp: StackAddress) {
```
Comment (vext01) [inline src/gc.rs]: Why are we placing ABI constraints here? Does C code call this?

File: src/gc.rs (hunk 10 above)
Comment (jacob-hughes) [reply src/gc.rs]: That makes sense,  but in this instance, collection can never happen when the collector is in any state other than `CollectorState::Ready`, even if new variants are added.

File: src/lib.rs (hunk 11 above)
Comment (vext01) [reply src/lib.rs]: As long as there are no aliasing problems... I have no idea.

File: src/gc.rs (hunk 23 above)
Comment (jacob-hughes) [reply src/gc.rs]: Because this function is called from the `SpillRegisters_x86_64` stub which sets up its arguments according to SYS V.

File: src/lib.rs (hunk 21 above)
Comment (vext01) [reply src/lib.rs]: Did you try without? It may not actually deref.

File: src/lib.rs (hunk 13 above)
Comment (ltratt) [reply src/lib.rs]: OK, if that's the constraint then stating it explicitly is a good idea. The explanation doesn't have to be long, so long as it's clear.

File: src/lib.rs (hunk 21 above)
Comment (jacob-hughes) [reply src/lib.rs]: In Rust you have to compose both operators if you want to get a reference from a raw pointer. It doesn't deref now, but the references are always considered dereference-able, so if you construct one from a raw pointer you better be sure that it's safe to de-reference it in the future.

Comment (vext01) [reply src/lib.rs]: Right, but that didn't answer my question. Did you try it without?

(You are probably right)

Comment (jacob-hughes) [reply src/lib.rs]: I'm not sure what you mean? `&(self.objptr as *const T)` is invalid Rust code.

Comment (ltratt) [reply src/lib.rs]: Turning a raw pointer into a Rust reference is always unsafe.

File: src/gc.rs (hunk 10 above)
Comment (ltratt) [reply src/gc.rs]: Yeah, in this situation `_` did feel like it was reasonable to me (I explicitly thought about it!).

File: src/lib.rs (hunk 21 above)
Comment (vext01) [reply src/lib.rs]: > Turning a raw pointer into a Rust reference is always unsafe.

ok

Comment (jacob-hughes) [reply src/lib.rs]: @vext01 Sorry, I understand now that you meant try and remove the unsafe block, not the `*`.

File: src/gc.rs (hunk 10 above)
Comment (vext01) [reply src/gc.rs]: It's up to you two. Seemed like a good candidate for the rule to me (few variants, not inconceivable that a new GC state be added later).

File: src/lib.rs (hunk 13 above)
Comment (jacob-hughes) [reply src/lib.rs]: I've decided to get rid of `is_white` altogether to avoid confusion, and warn on using the inverse of `is_black`.

File: gc_tests/tests/on_stack_root_marking.rs (hunk 22 above)
Comment (jacob-hughes) [reply gc_tests/tests/on_stack_root_marking.rs]: Fixed

File: src/lib.rs (hunk 19 above)
Comment (jacob-hughes) [reply src/lib.rs]: Fixed

**Comment by vext01** [comment]
LGTM

File: src/lib.rs [hunk 24]
```diff
@@ -179,6 +179,14 @@ pub fn collect() {
 pub struct Debug;
 
 impl Debug {
+    /// Checks that the object was marked as reachable in the last collection.
```
Comment (ltratt) [inline src/lib.rs]: It's a bit difficult to know what true/false means here. How about saying `Returns `true` if this object was reachable in the last collection`?

Comment (jacob-hughes) [reply src/lib.rs]: Sure

**Comment by jacob-hughes** [comment]
Ready for re-review

**Comment by ltratt** [comment]
Please squash.

**Comment by jacob-hughes** [comment]
Squashed

**Comment by ltratt** [comment]
Is " Add single API to interact with allocation block metadata " supposed to be "Add *simple* API to ..."?

In https://github.com/softdevteam/gcmalloc/pull/6/commits/0a16bd587fa123fc7dbd70d04adffea92afa270e "not in anyway" -> "not in any way".

Feel free to force push updates.

**Comment by jacob-hughes** [comment]
Yes, not quite sure how I missed that one. Fixed
